### PR TITLE
exposed ttl for spf cache work

### PR DESCRIPTION
**Pull request by csucu**
part of the work for redsift/ipforensics-svc#62

### CODE REVIEW

**Review by amnonbc — COMMENTED** [review]
we should add a circleci job to run the tests, and remove the unused travis file.

File: resolver_limited.go [hunk 1]
```diff
@@ -32,7 +33,7 @@ func (r *LimitedResolver) canLookup() bool {
 
 // LookupTXT returns the DNS TXT records for the given domain name.
 // Used for "exp" modifier and do not cause DNS query.
```
Comment (dmotylev) [inline resolver_limited.go]: TTL is not documented

File: resolver_limited.go [hunk 2]
```diff
@@ -41,9 +42,9 @@ func (r *LimitedResolver) LookupTXT(name string) ([]string, error) {
 // by underlying resolver exceed the limit.
 // It will also return ErrDNSPermerror upon DNS call return error NXDOMAIN
 // (RCODE 3)
-func (r *LimitedResolver) LookupTXTStrict(name string) ([]string, error) {
+func (r *LimitedResolver) LookupTXTStrict(name string) ([]string, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_limited.go:45]: TTL is not documented

File: resolver_limited.go [hunk 3]
```diff
@@ -66,9 +67,9 @@ func (r *LimitedResolver) Exists(name string) (bool, error) {
 // If any address matches, the mechanism matches
 // Returns false and ErrDNSLimitExceeded if total number of lookups made
 // by underlying resolver exceed the limit.
-func (r *LimitedResolver) MatchIP(name string, matcher IPMatcherFunc) (bool, error) {
+func (r *LimitedResolver) MatchIP(name string, matcher IPMatcherFunc) (bool, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_limited.go:70]: TTL is not documented

File: resolver_limited.go [hunk 4]
```diff
@@ -85,9 +86,9 @@ func (r *LimitedResolver) MatchIP(name string, matcher IPMatcherFunc) (bool, err
 //
 // Returns false and ErrDNSLimitExceeded if total number of lookups made
 // by underlying resolver exceed the limit.
-func (r *LimitedResolver) MatchMX(name string, matcher IPMatcherFunc) (bool, error) {
+func (r *LimitedResolver) MatchMX(name string, matcher IPMatcherFunc) (bool, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_limited.go:89]: TTL is not documented

File: resolver_miekg.go [hunk 5]
```diff
@@ -156,47 +156,67 @@ func (r *miekgDNSResolver) exchange(req *dns.Msg) (*dns.Msg, error) {
 }
 
 // LookupTXT returns the DNS TXT records for the given domain name.
-func (r *miekgDNSResolver) LookupTXT(name string) ([]string, error) {
+func (r *miekgDNSResolver) LookupTXT(name string) ([]string, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_miekg.go:160]: TTL is not documented

File: resolver_miekg.go [hunk 6]
```diff
@@ -156,47 +156,67 @@ func (r *miekgDNSResolver) exchange(req *dns.Msg) (*dns.Msg, error) {
...
+		ttl = 0
+	}
+
+	return txts, time.Duration(ttl) * time.Second, nil
 }
 
 // LookupTXTStrict returns DNS TXT records for the given name, however it
 // will return ErrDNSPermerror upon NXDOMAIN (RCODE 3)
-func (r *miekgDNSResolver) LookupTXTStrict(name string) ([]string, error) {
+func (r *miekgDNSResolver) LookupTXTStrict(name string) ([]string, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_miekg.go:190]: TTL is not documented

File: resolver_miekg.go [hunk 7]
```diff
@@ -214,7 +234,26 @@ func (r *miekgDNSResolver) Exists(name string) (bool, error) {
...
+			if d := a.Header().Ttl; d < ttl {
+				ttl = d
+			}
+		}
+	}
+
+	if len(txts) == 0 {
+		ttl = 0
+	}
+*/
```
Comment (dmotylev) [inline resolver_miekg.go]: This comment seems lost. Could be removed?

File: resolver_miekg.go [hunk 8]
```diff
@@ -225,18 +264,23 @@ func matchIP(rrs []dns.RR, matcher IPMatcherFunc, name string) (bool, error) {
...
-	return false, nil
+	return false, 0, nil
 }
 
 // MatchIP provides an address lookup, which should be done on the name
 // using the type of lookup (A or AAAA).
 // Then IPMatcherFunc used to compare checked IP to the returned address(es).
 // If any address matches, the mechanism matches
-func (r *miekgDNSResolver) MatchIP(name string, matcher IPMatcherFunc) (bool, error) {
+func (r *miekgDNSResolver) MatchIP(name string, matcher IPMatcherFunc) (bool, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_miekg.go:283]: TTL is not documented

File: resolver_miekg.go [hunk 9]
```diff
@@ -274,24 +318,24 @@ func (r *miekgDNSResolver) MatchIP(name string, matcher IPMatcherFunc) (bool, er
...
-	return false, nil
+	return false, 0, nil
 }
 
 // MatchMX is similar to MatchIP but first performs an MX lookup on the
 // name.  Then it performs an address lookup on each MX name returned.
 // Then IPMatcherFunc used to compare checked IP to the returned address(es).
 // If any address matches, the mechanism matches
-func (r *miekgDNSResolver) MatchMX(name string, matcher IPMatcherFunc) (bool, error) {
+func (r *miekgDNSResolver) MatchMX(name string, matcher IPMatcherFunc) (bool, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_miekg.go:332]: TTL is not documented

File: resolver_retry.go [hunk 10]
```diff
@@ -67,27 +67,27 @@ func NewRetryResolver(rr []Resolver, opts ...RetryResolverOption) Resolver {
...
+				return v, ttl, err
 			}
 		}
 		time.Sleep(r.backoff(attempt))
 	}
 }
 
 // LookupTXT returns the DNS TXT records for the given domain name.
-func (r *retryResolver) LookupTXT(name string) ([]string, error) {
+func (r *retryResolver) LookupTXT(name string) ([]string, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_retry.go:84]: TTL is not documented

File: resolver_retry.go [hunk 11]
```diff
@@ -114,13 +114,13 @@ func (r *retryResolver) Exists(name string) (bool, error) {
 // using the type of lookup (A or AAAA).
 // Then IPMatcherFunc used to compare checked IP to the returned address(es).
 // If any address matches, the mechanism matches
-func (r *retryResolver) MatchIP(name string, matcher IPMatcherFunc) (bool, error) {
+func (r *retryResolver) MatchIP(name string, matcher IPMatcherFunc) (bool, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_retry.go:117]: TTL is not documented

File: resolver_retry.go [hunk 12]
```diff
@@ -131,13 +131,13 @@ func (r *retryResolver) MatchIP(name string, matcher IPMatcherFunc) (bool, error
 // name.  Then it performs an address lookup on each MX name returned.
 // Then IPMatcherFunc used to compare checked IP to the returned address(es).
 // If any address matches, the mechanism matches
-func (r *retryResolver) MatchMX(name string, matcher IPMatcherFunc) (bool, error) {
+func (r *retryResolver) MatchMX(name string, matcher IPMatcherFunc) (bool, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_retry.go:134]: TTL is not documented

File: resolver_std.go [hunk 13]
```diff
@@ -33,7 +34,7 @@ func errDNS(e error) error {
 
 // LookupTXTStrict returns DNS TXT records for the given name, however it
 // will return ErrDNSPermerror upon NXDOMAIN (RCODE 3)
-func (r *DNSResolver) LookupTXTStrict(name string) ([]string, error) {
+func (r *DNSResolver) LookupTXTStrict(name string) ([]string, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_std.go:37]: TTL is not documented

File: resolver_std.go [hunk 14]
```diff
@@ -49,25 +50,25 @@ func (r *DNSResolver) LookupTXTStrict(name string) ([]string, error) {
...
-		return nil, err
+		return nil, 0, err
 	}
-	return txts, nil
+	return txts, 0, nil
 }
 
 // LookupTXT returns the DNS TXT records for the given domain name.
-func (r *DNSResolver) LookupTXT(name string) ([]string, error) {
+func (r *DNSResolver) LookupTXT(name string) ([]string, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_std.go:65]: TTL is not documented

File: resolver_std.go [hunk 15]
```diff
@@ -84,36 +85,37 @@ func (r *DNSResolver) Exists(name string) (bool, error) {
...
+	ttl   time.Duration
 	err   error
 }
 
 // MatchIP provides an address lookup, which should be done on the name
 // using the type of lookup (A or AAAA).
 // Then IPMatcherFunc used to compare checked IP to the returned address(es).
 // If any address matches, the mechanism matches
-func (r *DNSResolver) MatchIP(name string, matcher IPMatcherFunc) (bool, error) {
+func (r *DNSResolver) MatchIP(name string, matcher IPMatcherFunc) (bool, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_std.go:96]: TTL is not documented

File: resolver_std.go [hunk 16]
```diff
@@ -84,36 +85,37 @@ func (r *DNSResolver) Exists(name string) (bool, error) {
...
-	return false, nil
+	return false, 0, nil
 }
 
 // MatchMX is similar to MatchIP but first performs an MX lookup on the
 // name.  Then it performs an address lookup on each MX name returned.
 // Then IPMatcherFunc used to compare checked IP to the returned address(es).
 // If any address matches, the mechanism matches
-func (r *DNSResolver) MatchMX(name string, matcher IPMatcherFunc) (bool, error) {
+func (r *DNSResolver) MatchMX(name string, matcher IPMatcherFunc) (bool, time.Duration, error) {
```
Comment (dmotylev) [inline resolver_std.go:114]: TTL is not documented

File: spf.go [hunk 17]
```diff
@@ -60,10 +60,10 @@ type IPMatcherFunc func(ip net.IP, name string) (bool, error)
 // Resolver provides abstraction for DNS layer
 type Resolver interface {
```
Comment (dmotylev) [inline spf.go:61]: The new output parameter TTL is not documented on the interface methods

**Review by dmotylev — CHANGES_REQUESTED** [review]

**Comment by dmotylev** [comment]
No TTL returned for 
https://github.com/redsift/spf/blob/e0770558e23fca2a39b0f8b9119769abe7d8bcc9/resolver_miekg.go#L225

File: spf.go (hunk 17 above)
Comment (csucu) [reply spf.go:61]: done

File: resolver_std.go (hunk 16 above)
Comment (csucu) [reply resolver_std.go:114]: done

File: resolver_std.go (hunk 15 above)
Comment (csucu) [reply resolver_std.go:96]: done

File: resolver_std.go (hunk 14 above)
Comment (csucu) [reply resolver_std.go:65]: done

File: resolver_std.go (hunk 13 above)
Comment (csucu) [reply resolver_std.go:37]: done

File: resolver_miekg.go (hunk 5 above)
Comment (csucu) [reply resolver_miekg.go:160]: done

File: resolver_limited.go (hunk 4 above)
Comment (csucu) [reply resolver_limited.go:89]: done

File: resolver_limited.go (hunk 3 above)
Comment (csucu) [reply resolver_limited.go:70]: done

File: resolver_limited.go (hunk 2 above)
Comment (csucu) [reply resolver_limited.go:45]: done

File: resolver_limited.go (hunk 1 above)
Comment (csucu) [reply resolver_limited.go]: done

**Review by dmotylev — APPROVED** [review]

**Pull request merged**